events.go

```diff
@@ -0,0 +1,182 @@
+type ChallengeResponse struct {
+	Challenge string
+}
+
+// EventsAPICallbackEvent is the main EventsAPI event.
+type EventsAPICallbackEvent struct {
+	Type        string           `json:"type"`
+	Token       string           `json:"token"`
+	TeamID      string           `json:"team_id"`
+	APIAppID    string           `json:"api_app_id"`
+	InnerEvent  *json.RawMessage `json:"event"`
+	AuthedUsers []string         `json:"authed_users"`
+	EventID     string           `json:"event_id"`
+	EventTime   int              `json:"event_time"`
+}
+
+// AppMentionEvent is an EventsAPI subscribable event.
+type AppMentionEvent struct {
+	Type           string      `json:"type"`
+	User           string      `json:"user"`
+	Text           string      `json:"text"`
+	TimeStamp      string      `json:"ts"`
+	Channel        string      `json:"channel"`
+	EventTimeStamp json.Number `json:"event_ts"`
+}
+
+func parseOuterEvent(rawE json.RawMessage) EventsAPIEvent {
+	e := &Event{}
+	err := json.Unmarshal(rawE, e)
+	if err != nil {
+		return EventsAPIEvent{
+			"unmarshalling_error",
+			&UnmarshallingErrorEvent{err},
+			EventsAPIInnerEvent{},
+		}
+	}
+	if e.Type == CallbackEvent {
+		cbEvent := &EventsAPICallbackEvent{}
+		err = json.Unmarshal(rawE, cbEvent)
+		if err != nil {
+			return EventsAPIEvent{
+				"unmarshalling_error",
+				&UnmarshallingErrorEvent{err},
+				EventsAPIInnerEvent{},
+			}
+		}
+		return EventsAPIEvent{
+			e.Type,
+			cbEvent,
+			EventsAPIInnerEvent{},
+		}
+	}
+	urlVE := &EventsAPIURLVerificationEvent{}
+	err = json.Unmarshal(rawE, urlVE)
+	if err != nil {
+		return EventsAPIEvent{
+			"unmarshalling_error",
+			&UnmarshallingErrorEvent{err},
+			EventsAPIInnerEvent{},
+		}
+	}
+	return EventsAPIEvent{
+		e.Type,
+		urlVE,
+		EventsAPIInnerEvent{},
+	}
+}
+
+func parseInnerEvent(e *EventsAPICallbackEvent) (EventsAPIEvent, error) {
+	iE := &Event{}
+	rawInnerJSON := e.InnerEvent
+	err := json.Unmarshal(*rawInnerJSON, iE)
+	if err != nil {
+		return EventsAPIEvent{
+			"unmarshalling_error",
+			&UnmarshallingErrorEvent{err},
+			EventsAPIInnerEvent{},
+		}, err
+	}
+	v, exists := EventMapping[iE.Type]
+	if !exists {
+		return EventsAPIEvent{
+			iE.Type,
+			nil,
+			EventsAPIInnerEvent{},
+		}, fmt.Errorf("Inner Event does not exist! %s", iE.Type)
+	}
+	t := reflect.TypeOf(v)
+	recvEvent := reflect.New(t).Interface()
+	err = json.Unmarshal(*rawInnerJSON, recvEvent)
+	if err != nil {
+		return EventsAPIEvent{
+			"unmarshalling_error",
+			&UnmarshallingErrorEvent{err},
+			EventsAPIInnerEvent{},
+		}, err
+	}
+	return EventsAPIEvent{
+		e.Type,
+		e,
+		EventsAPIInnerEvent{iE.Type, recvEvent},
+	}, nil
+}
+
+// ParseEventsAPIEvent parses the outter and inner events (if applicable) of an events
+// api event returning a EventsAPIEvent type. If the event is a url_verification event,
+// the inner event is empty.
+func (api *Client) ParseEventsAPIEvent(rawEvent json.RawMessage) (EventsAPIEvent, error) {
```

Comment:
this doesn't need to be attached to the client. i'd just remove the debug as it'll return an error in that case.

I'd also make a custom error type for the events API to have the metadata in a structured format and for the `Error() string` method just return whatever slack claims the error is unaltered.